src/compiler/checker.ts

```diff
@@ -21637,7 +21637,8 @@ namespace ts {
 
         function getOptionalType(type: Type, isProperty = false): Type {
             Debug.assert(strictNullChecks);
-            return type.flags & TypeFlags.Undefined ? type : getUnionType([type, isProperty ? missingType : undefinedType]);
+            const missingOrUndefined = isProperty ? missingType : undefinedType;
+            return type.flags & TypeFlags.Undefined || type.flags & TypeFlags.Union && (type as UnionType).types[0] === missingOrUndefined ? type : getUnionType([type, missingOrUndefined]);

```

Comment:
The `types[0]` check always seems extremely fragile, but I suppose it's convenient.

Reply:
Yeah, but we make that assumption in several other places, so seems fine to do it here as well.